src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,371 @@
+                .add(HttpUtil.CONTENT_TYPE_HEADER, HttpHeaderValues.APPLICATION_JSON);
+
+        final Future<HttpClientResponse> responseFuture = httpClient.post(
+                greenbidsAnalyticsProperties.getAnalyticsServerUrl(),
+                headers,
+                commonMessageJson,
+                greenbidsAnalyticsProperties.getTimeoutMs());
+
+        return responseFuture.compose(this::processAnalyticServerResponse);
+    }
+
+    private GreenbidsPrebidExt parseBidRequestExt(BidRequest bidRequest) {
+        return Optional.ofNullable(bidRequest)
+                .map(BidRequest::getExt)
+                .map(ExtRequest::getPrebid)
+                .map(ExtRequestPrebid::getAnalytics)
+                .filter(this::isNotEmptyObjectNode)
+                .map(analytics -> (ObjectNode) analytics.get(BID_REQUEST_ANALYTICS_EXTENSION_NAME))
+                .map(this::toGreenbidsPrebidExt)
+                .orElse(GreenbidsPrebidExt.builder().build());
+    }
+
+    private GreenbidsPrebidExt toGreenbidsPrebidExt(ObjectNode adapterNode) {
+        try {
+            return jacksonMapper.mapper().treeToValue(adapterNode, GreenbidsPrebidExt.class);
+        } catch (JsonProcessingException e) {
+            throw new PreBidException("Error decoding bid request analytics extension: " + e.getMessage(), e);
+        }
+    }
+
+    private boolean isNotEmptyObjectNode(JsonNode analytics) {
+        return analytics != null && analytics.isObject() && !analytics.isEmpty();
+    }
+
+    private Future<Void> processAnalyticServerResponse(HttpClientResponse response) {
+        final int responseStatusCode = response.getStatusCode();
+        if (responseStatusCode >= 200 && responseStatusCode < 300) {
+            return Future.succeededFuture();
+        }
+        return Future.failedFuture(new PreBidException("Unexpected response status: " + response.getStatusCode()));
+    }
+
+    private boolean isSampled(double samplingRate, String greenbidsId) {
+        if (samplingRate < 0 || samplingRate > 1) {
+            logger.warn("Warning: Sampling rate must be between 0 and 1");
+            return false;
+        }
+
+        final double exploratorySamplingRate = samplingRate
+                * greenbidsAnalyticsProperties.getExploratorySamplingSplit();
+        final double throttledSamplingRate = samplingRate
+                * (1.0 - greenbidsAnalyticsProperties.getExploratorySamplingSplit());
+
+        final int hashInt = Integer.parseInt(
+                greenbidsId.substring(greenbidsId.length() - 4), 16);
+        final boolean isPrimarySampled = hashInt < exploratorySamplingRate * RANGE_16_BIT_INTEGER_DIVISION_BASIS;
+        final boolean isExtraSampledOutOfExploration = hashInt >= (1 - throttledSamplingRate)
+                * RANGE_16_BIT_INTEGER_DIVISION_BASIS;
+
+        return isPrimarySampled || isExtraSampledOutOfExploration;
+    }
+
+    private CommonMessage createBidMessage(
```

Comment:
This method is too large, please extract some inner logic to separate methods

Reply:
Have extracte `getSeatsWithBids` and `getSeatsWithNonBids` to separate methods 👍